### DISCUSSION TITLE
Using the LPS with a Pixhawk

### DISCUSSION THREAD

**Discussion by magicbycalvin** [bitcraze/discussions#87] [Q&A]
Hey folks,

My lab is interested in using the UWB capabilities from the Loco Positioning System with Pixhawk flight controllers for a large scale experiment this Summer. Due to the global supply chain issues, we are restricted to the hardware we have on hand. Otherwise, we would have purchased several of the CF Bolt and Roadrunner products. The end goal is to have one large robot whose absolute position is known and localize all other robots off of that one using UWB (this is one of the areas of research we plan on figuring out down the line since currently the LPS uses fixed, known locations for the anchors).

The first step we would like to accomplish is simply localizing a robot using a Pixhawk controller with the standard LPS. We have access to 16 LPS nodes and several LPS decks. Has there been any work in getting one of those two products to communicate with a Pixhawk or other microcontroller? If not, would you kindly point me in the right direction for where I can begin researching and eventually implementing firmware changes to allow the node and/or deck to communicate with a Pixhawk?

Once the above proof of concept has been accomplished, then we will move towards localization of a team of mobile vehicles. To accomplish this, we will need access to the raw LPS measurements. Hopefully if the above challenge of integrating an LPS node/deck with the Pixhawk is solved, this problem should also be solved. But if not, is there any existing code that can be used to send this data to the Pixhawk?

Finally, are there any other considerations I may be missing? And if you have any other resources you can share to help figure this out both theory and implementation-wise, I would really appreciate it.

Thank you for your time!

**Comment by krichardsson** [comment]
Hi!

The Bolt will be available again in our store today or tomorrow if this solves any of your problems :-)

I'm not aware of any pixhawk implementations of Loco functionality, but I might be wrong. 
The Crazyflie side of the Loco functionality is implemented in a few different places, I will base the pointers on the assumption that you will use TDoA3
1. The low level drivers for the DWM1000 can be found in [bitcraze/libdw1000 ](https://github.com/bitcraze/libdw1000)
2. The communication layer is implemented in https://github.com/bitcraze/crazyflie-firmware/blob/master/src/deck/drivers/src/lpsTdoa3Tag.c
3. The TDoA functionality is implemented in the TDoA engine module https://github.com/bitcraze/crazyflie-firmware/tree/master/src/utils/src/tdoa
4. The position estimation is implemented in the kalman estimator

If you want to do inter robot ranging there is an old issue where this has been discussed a bit, bitcraze/crazyflie-firmware#576. We have not implemented this yet in the official firmware but there are some research groups that have done there own implementations or modified my sketch (see the issue). This is perhaps something that we should clean up and make a proper implementation of, let us know if you will use something like this.

**Comment by Bjarne33** [comment]
Hey, 

I just saw your post. I used the Loco Positioning System and the Roadrunner to locate a ground robot, which uses the pixhawk. I programmed an application that logs the positioning data from the roadrunner and sends it over the serial port. I than modified the AP_Beacon library of the Ardupilot software, which runs on my pixhawk. I added a cpp file and a header file. You can have a look on the AP_Beacon_Pozyx file and documentation, it works in a similar way.

Good luck with your project!

**Reply by knmcguire** [reply]
Just mind that the roadrunner is not sold anymore, but a [bolt ](https://store.bitcraze.io/collections/kits/products/crazyflie-bolt-1-1)+ [loco deck](https://store.bitcraze.io/collections/decks/products/loco-positioning-deck) combo is practically doing the same thing.
